#3783 Lookup patients online when vaccine dispensing

joshxg opened the pull request:
Fixes #3783 

## Change summary

- Adds online lookup to vaccine dispensing
- Lots of side effects !

## Testing

- [ ] Can vaccine dispense to an existing patient from local storage
- [ ] can vaccine dispense to a patient on the line
- [ ] can vaccine dispense to a new patient created
- [ ] Can create a new patient in normal dispensing
- [ ] Can create a patient in normal dispensing

### Related areas to think about

- Normal dispensing edit/creation
- Vaccine dispensing edit/creation
- Errors in communicating with the server

joshxg commented on src/actions/Entities/NameActions.js:
```diff
@@ -11,9 +11,10 @@ export const NAME_ACTIONS = {
 };
 
-const createDefaultName = (type = 'patient', id) => ({
+export const createDefaultName = (type = 'patient', id) => ({
```
rather than passing an id in and creating one here, started to create a default one in components and pass the patient.

This is because I would be passing a patient when selecting one from having been lookedup.

Just occurs to me I mightv'e been able to just quickly save the patient when selecting and made a lot less changes 🤔

joshxg commented on src/actions/Entities/NameActions.js:
```diff
@@ -42,13 +43,12 @@ const update = (id, field, value) => ({
   batch(() => {
     Object.entries(detailsEntered).forEach(([key, value]) => {
-      if (currentKeys.includes(key) && currentPatient[key] !== value) {
+      if (currentPatient[key] !== value) {
```
wasn't sure what this was really trying to do - but it stopped me from updating when for example I started with a 'default name' record and went to edit values like gender which weren't in the default name yet

joshxg commented on src/actions/Entities/NameNoteActions.js:
```diff
@@ -43,12 +44,29 @@ const createDefaultNameNote = (nameID = '') => {
+  );
+  const [mostRecentPCD] = sorted;
+  return mostRecentPCD;
+};
```
This is changed such that it can look up the most recent name note for an existing patient or one from the lookup API.

This will need this mSupply PR https://github.com/sussol/msupply/pull/8280 to work- I haven't tested it with it yet

joshxg commented on src/actions/PatientActions.js:
```diff
@@ -18,7 +18,7 @@ export const PATIENT_ACTIONS = {
 
 const closeModal = () => ({ type: PATIENT_ACTIONS.COMPLETE });
-const createPatient = () => ({ type: PATIENT_ACTIONS.PATIENT_CREATION });
+const createPatient = patient => ({ type: PATIENT_ACTIONS.PATIENT_CREATION, payload: { patient } });
```
This is for creating a patient in normal dispensing (because I guess I thought it was a good idea to have name actions and patient actions 🤔 )

This is setting a 'default name' as the one being edited now rather than sorting it out when saving, so that there is a stable ID to use for the name note.

joshxg commented on src/database/DataTypes/Name.js:
```diff
@@ -161,6 +161,7 @@ export class Name extends Realm.Object {
       thisStoresPatient: this.thisStoresPatient,
       nationality: this.nationality?.toJSON(),
       ethnicity: this.ethnicity?.toJSON(),
+      nameNotes: this.nameNotes?.map(nameNote => nameNote.toObject()),
```
This is so the names from realm convert to an object which more closely resembles the ones sent from the patient lookup api

joshxg commented on src/database/DataTypes/NameNote.js:
```diff
@@ -15,6 +15,10 @@ export class NameNote extends Realm.Object {
 
+  get patientEventID() {
+    return this.patientEvent?.id;
+  }
```
Same as above- so the name note is more similar to those which will be sent from the api

joshxg commented on src/hooks/useLocalAndRemotePatients.js:
```diff
@@ -0,0 +1,153 @@
+import { useEffect, useReducer } from 'react';
```
Aaaand this is where it gets a bit hairy.

I started off using thunks and redux to do the API calls.

It was getting pretty messy, knowing when data was coming from APIs or from realm, to know when to filter. It was also hard to track loading/error states through redux.

I thought to use the `useFetch` hook which handles most of this already.

What I ended up with was this hook which is an attempt to create an interface where data from two different sources are kind of merged to form a single source of state.

There will also need to be pagination logic added to this.

It's a bit complex but I think it is less so then mixing hooks and different redux stuff going on

joshxg commented on src/pages/DispensingPage.js:
```diff
@@ -349,10 +350,16 @@ const mapDispatchToProps = dispatch => ({
   editPatient: patient =>
     batch(() => {
-      dispatch(NameNoteActions.createSurveyNameNote(patient));
+      dispatch(NameNoteActions.createSurveyNameNote(UIDatabase.get('Name', patient)));
```
`patient` was actually `patientID`- and `createSurveyNameNote` was changed to accept a patient, not the ID

joshxg commented on src/pages/DispensingPage.js:
```diff
@@ -349,10 +350,16 @@ const mapDispatchToProps = dispatch => ({
+    batch(() => {
+      const patient = createDefaultName('patient', generateUUID());
+      dispatch(PatientActions.createPatient(patient));
+      dispatch(NameNoteActions.createSurveyNameNote(patient));
```
Creating these now with a stable patient ref

joshxg commented on src/reducers/Entities/NameReducer.js:
```diff
@@ -42,7 +42,11 @@ export const NameReducer = (state = initialState(), action) => {
+      // patient API.
+      const asPlainObject = name.toJSON ? name.toJSON() : name;
+
+      return { ...state, editing: asPlainObject };
```
this can now either be from the api or from realm

joshxg commented on src/widgets/Tabs/PatientSelect.js:
```diff
@@ -54,22 +61,78 @@ import { Paper } from '../Paper';
   const handleUpdate = (key, value) => {
     updateForm(key, value);
-    debouncedFilter(key, value);
+    filter({ ...completedForm, [key]: value });
```
need to do this here as completedForm is not correct until the above `updateForm` has been dispatched

joshxg commented on src/widgets/modalChildren/PatientEditModal.js:
```diff
@@ -22,62 +21,53 @@ export const PatientEditModalComponent = ({
-            textStyle={styles.cancelButtonTextStyle}
-            text={modalStrings.cancel}
-          />
+}) => (
```
don't need the `onOpen` now, because it is creating the required records in `DispensingPage` callbacks

joshxg commented on src/widgets/modalChildren/PatientEditModal.js:
```diff
@@ -123,26 +113,24 @@ PatientEditModalComponent.propTypes = {
-    onSave({ id, ...completedForm });
-    if (surveySchema) onSaveSurvey(id);
+    onSave(completedForm);
+    if (surveySchema) onSaveSurvey();
```
don't need to use these ids now because `dispensingPage` callbacks are creating a default `patient` which has an `id`.

mark-prins commented:
@joshxg If I enter a DOB on the search form, then I get the response 'there was a problem communicating with the server'

mark-prins commented:
Found another problem - searching using DOB doesn't work: enter a DOB and I'm told 'Enter the patients details...'
Entering a first name and DOB will find no patients.

mark-prins reviewed (approved):
👍 
Well, that's awesome. Genius to use a reducer in a hook to handle it all.
There's a couple of issues with DOBs - can push on an merge this one and fix those in another issue if that's helpful though.

I'm slightly confused about the merging of results.. the reducer is setting rather than merging. The API lookup returns all patients, so there isn't a merge, just a list from the server. some of them exist locally and some don't 🤷‍♂️

mark-prins commented on src/localization/generalStrings.json:
```diff
@@ -1,5 +1,8 @@
 {
   "gb": {
+    "enter_patient_details": "Enter the patients details to start searching for them!",
```
:nit 
```suggestion
    "enter_patient_details": "Enter the patient's details to start searching for them!",
```
The details belong to the patient, so there should be an apostrophe, I think.

joshxg commented on src/database/DataTypes/Name.js:
```diff
@@ -161,6 +161,7 @@ export class Name extends Realm.Object {
       thisStoresPatient: this.thisStoresPatient,
       nationality: this.nationality?.toJSON(),
```
Oh it's on the realm class which each of our models extends.

I should probably rename all our overrides to `toObject` like I had planned but keep forgetting!

joshxg commented:
- Changed all the strings: 19f3749, 78908919dc47e6059185b6e16109b76b0d08fda6
- Added sorting: c5b73e96c47c4feb676fe329561ac1bf62126dba ... e412e59a3ceab6237b48b970de7b0772cfc0af14

- Re DOB- 67472940a195e7b08264471b982b3f02a234917b - I was so hung up on handling errors when the server wasn't there I didn't handle them correctly when it is there! I hadn't realised that when there are no results, the response from `useFetch` was `null` and there was also an error set. I don't particularly like using the error message as it's pretty error prone, but changing the behaviour for not throwing an error when the result set is empty is a bit messy, so just went with this...  Also, when the DOB was invalid an action wasn't properly dispatched. Hopefully adding that all together it has more consistent behaviour now!
- Re 'merge' : wrong word, I think. I am trying to describe the 'funnelling' of two different 'things' going on- local state and actions with remote state/data and the async actions being 'funnelled' into a single 'state', where just reading 'loading' of this state will tell you if the local db is searching or the async request is being sent and you don't have to read from two different arrays for patients, dependent on whether the last request was through the api or local storage. I think i'm making it seem more confusing than it is...

joshxg commented:
Going to merge this now.

I did notice a couple other things which I will make new issues for:
- Disable search online when in the middle of another search
- The debouncing in `FormControl` is annoying
- Update comments to remove 'merge'?
-

mark-prins commented:
thanks - I get it now :) I had in mind that you'd have a single array of objects, some from the api and some from realm; which would be awful. On re-reading the comments, they're fine. No need to change those.
Have to confess that I didn't even notice the button not disabling while searching - that's usually a thing that bugs me. The search was quick ( all running locally 😉 ) and everything was so smooth - didn't even think to click the button twice.

joshxg commented:
Ha, yeah- I think mine was just running a bit slower because I was accidentally running on mobile data 🤣
